src/core/server/server.cc

```diff
@@ -84,6 +86,227 @@
+}
+
+//
+// Server::ListenerInterface
+//
+
+void Server::ListenerWrapper::Start() {
+  if (IsServerListenerEnabled()) {
+    if (server_->config_fetcher() != nullptr) {
+      auto watcher =
+          std::make_unique<ListenerWrapper::ConfigFetcherWatcher>(this);
+      config_fetcher_watcher_ = watcher.get();
+      server_->config_fetcher()->StartWatch(
+          grpc_sockaddr_to_string(listener_->resolved_address(), false).value(),
+          std::move(watcher));
+    } else {
+      {
+        MutexLock lock(&mu_);
+        started_ = true;
+        is_serving_ = true;
+      }
+      listener_->Start();
+    }
+  } else {
+    listener_->Start();
+  }
+}
+
+void Server::ListenerWrapper::Stop() {
+  if (IsServerListenerEnabled()) {
+    absl::flat_hash_set<OrphanablePtr<ListenerInterface::LogicalConnection>>
+        connections;
+    {
+      MutexLock lock(&mu_);
+      // Orphan the connections so that they can start cleaning up.
+      connections = std::move(connections_);
+      is_serving_ = false;
+    }
+    if (config_fetcher_watcher_ != nullptr) {
+      CHECK_NE(server_->config_fetcher(), nullptr);
+      server_->config_fetcher()->CancelWatch(config_fetcher_watcher_);
+    }
+  }
+  GRPC_CLOSURE_INIT(&destroy_done_, ListenerDestroyDone, server_,
+                    grpc_schedule_on_exec_ctx);
+  listener_->SetOnDestroyDone(&destroy_done_);
+  listener_.reset();
+}
+
+absl::optional<ChannelArgs> Server::ListenerWrapper::AddLogicalConnection(
+    OrphanablePtr<ListenerInterface::LogicalConnection> connection,
+    const ChannelArgs& args, grpc_endpoint* endpoint) {
+  RefCountedPtr<grpc_server_config_fetcher::ConnectionManager>
+      connection_manager;
+  {
+    MutexLock lock(&mu_);
+    if (!is_serving_) {
+      // Not serving
+      return absl::nullopt;
+    }
+    connection_manager = connection_manager_;
+  }
+  ChannelArgs new_args = args;
+  if (server_->config_fetcher() != nullptr) {
+    if (connection_manager == nullptr) {
+      // Connection manager not available
+      return absl::nullopt;
+    }
+    absl::StatusOr<ChannelArgs> args_result =
+        connection_manager->UpdateChannelArgsForConnection(new_args, endpoint);
+    if (!args_result.ok()) {
+      return absl::nullopt;
+    }
+    auto* server_credentials =
+        (*args_result).GetObject<grpc_server_credentials>();
+    if (server_credentials == nullptr) {
+      // Could not find server credentials
+      return absl::nullopt;
+    }
+    auto security_connector =
+        server_credentials->create_security_connector(*args_result);
+    if (security_connector == nullptr) {
+      // Unable to create secure server with credentials
+      return absl::nullopt;
+    }
+    new_args = (*args_result).SetObject(security_connector);
+  }
+  MutexLock lock(&mu_);
+  if (!is_serving_ || connection_manager != connection_manager_) {
+    // Not serving
+    return absl::nullopt;
+  }
+  connections_.emplace(std::move(connection));
```

Comment:
If we take any of the early return paths above, then we aren't going to save `connection` in the map, which means that we'll orphan the object as soon as it goes out of scope.  Doesn't that mean that whenever there's no config fetcher, we should shut down the handshake immediately?

I think I must be missing something, because it looks like the existing code has the same problem.  Why isn't this resulting in always shutting down new connections immediately in non-xDS cases?

Reply:
Thanks, that was definitely part of my confusion.  But I think the other part of my confusion here is that it seems odd that orphaning the `LogicalConnection` object doesn't actually shut down the connection, it only shuts down handshaking.  For example, in the case where there is no config fetcher, we will wind up removing the entry after handshaking completes, at which point we will call `Orphan()` -- which is currently a no-op, since the only thing `Orphan()` does is cancel the handshake, which at this point is already complete.  This seems a little counter-intuitive, since I'd expect `Orphan()` to initiate a full shutdown of the object -- i.e., in this case, I would expect it to shut down the connection.

I'd ideally like to clean this up somehow to make the structure a little more intuitive, but I'm not sure if there's a good way to do it.  I think this weirdness stems from the fact that the server code actually deals with connections very differently while the handshake is still going on than it does once it has a transport -- e.g., before handshaking is done, server shutdown triggers handshaker shutdown, but when there is a transport, server shutdown triggers sending a GOAWAY.  But even that isn't a very clear division, because the handshaking isn't truly complete until after we see the HTTP/2 SETTINGS frame, which is after the transport has already been created.  Given that, it's not clear to me how to restructure this to make this cleaner without a lot more work, but maybe you'll have some ideas here.

At minimum, let's at least document the weird expectation on the `LogicalConnection` object that `Orphan()` should cancel a handshake but not actually shut down the connection.

Reply:
Notes from our offline discussion:

One possible approach here would be to change the server to unconditionally track the `LogicalConnection` for its entire lifetime, even if there is no config fetcher, instead of separately storing the transports like it does today.  However, we'd need to carefully do this in a way that doesn't increase per-connection memory usage.

@ctiller may want to weigh in here as well, especially w.r.t. any possible related structural changes that he might have in mind in the server code for after the promise conversion is done.  Maybe as part of that, there might be ways to do this without increasing memory usage.